src/Messages/OWSMessageSender.m

```diff
@@ -628,11 +629,17 @@ - (void)sendSyncTranscriptForMessage:(TSOutgoingMessage *)message
             // DEPRECATED - Remove after all clients have been upgraded.
             BOOL isLegacyMessage = ![message isKindOfClass:[OWSOutgoingSyncMessage class]];
 
-            NSDictionary *messageDict = [self encryptedMessageWithPlaintext:plainText
-                                                                toRecipient:recipient.uniqueId
-                                                                   deviceId:deviceNumber
-                                                              keyingStorage:[TSStorageManager sharedManager]
-                                                                     legacy:isLegacyMessage];
+            __block NSDictionary *messageDict;
+            // Mutating session state is not thread safe, so we operate on a serial queue, shared with decryption
+            // operations.
+            dispatch_sync([OWSDispatch sessionCipher], ^{
```

Comment:
Yeah, OWSDispatch might be a better place.  But it would just be nice to have a certain understanding:

* We should avoid dispatch_sync, only when necessary.
* We should have broad rules about which threads and queues can block on each other (i.e. dispatch sync).